packages/core/src/storages/request_queue_v2.ts

```diff
@@ -154,13 +176,85 @@ export class RequestQueue extends RequestProvider {
         //    will not put the request again to queueHeadDict.
         if (request.handledAt) {
             this.log.debug('Request fetched from the beginning of queue was already handled', { nextRequestId });
-            this.recentlyHandledRequestsCache.add(nextRequestId, true);
             return null;
         }
 
         return request;
     }
 
+    /**
+     * @inheritDoc
+     */
+    override async isFinished(): Promise<boolean> {
+        if (this.inProgressRequestBatches.length > 0) {
+            return false;
+        }
+
+        if (this.queueHeadIds.length() > 0) {
+            return false;
+        }
```

Comment:
i guess the duplicity (same check 5 lines later) here is for performance reasons?

Reply:
Yup. If the `queueHeadIds` is non-empty, we return immediately, otherwise we try to fetch something from the upstream queue, which may take time.